### PR TITLE
Lucene 9.12 Upgrade

### PR DESCRIPTION
**Pull request by abhioncbr**
As per the [issue](https://github.com/apache/pinot/issues/14232), In this PR, we are upgrading the Lucene version to 9.12 and introducing the index files based on the newer version.

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
We cannot directly bump creator version to 9.12 because server might have not been upgraded and might not able to read the new file. We want to add 9.12 support to reader first, then change creator in the next release

**Review by Jackie-Jiang — COMMENTED** [review]
Can you also check why the compatibility test didn't fail? Does it cover text index?

It was failing initially, but with the support for both V99 and V912-based index read, it turned green.

**Comment by abhioncbr** [comment]
> We cannot directly bump creator version to 9.12 because server might have not been upgraded and might not able to read the new file. We want to add 9.12 support to reader first, then change creator in the next release

In my understanding, we can upgrade in one shot. We support the index read from both V99 and V912-based file extensions, but we write the index only with V912-based file extensions. That's the same approach we used last time as well([PR](https://github.com/apache/pinot/pull/12866) for reference).

**Comment by abhioncbr** [comment]
> Can you also check why the compatibility test didn't fail? Does it cover text index?

It was failing initially, but with the support for both V99 and V912-based index read, it turned green.

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/LuceneFSTIndexReader.java [hunk 1]
```diff
@@ -52,8 +51,7 @@ public LuceneFSTIndexReader(PinotDataBuffer pinotDataBuffer)
     _dataBufferIndexInput = new PinotBufferIndexInput(_dataBuffer, 0L, _dataBuffer.size());
 
     _readFST =
-        new FST<>(FST.readMetadata(_dataBufferIndexInput, PositiveIntOutputs.getSingleton()),
-            _dataBufferIndexInput, new OffHeapFSTStore());
+        new FST<>(FST.readMetadata(_dataBufferIndexInput, PositiveIntOutputs.getSingleton()), _dataBufferIndexInput);
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/LuceneFSTIndexReader.java:54]: In recent tests, I found OnHeapFSTStore got used by default
```
public FST(FSTMetadata<T> metadata, DataInput in) throws IOException {
    this(metadata, in, new OnHeapFSTStore(DEFAULT_MAX_BLOCK_BITS));
  }
```

I'm looking at how to create the reader with OffheapFSTStore? Feel free to chime in.

Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/LuceneFSTIndexReader.java:54]: https://github.com/apache/pinot/pull/14457

**Pull request merged**